### PR TITLE
refactor: refactor cluster package

### PR DESCRIPTION
**Pull request by chunshao90**
# Which issue does this PR close?

Refer to #94 

# Rationale for this change
 
 

# What changes are included in this PR?
Extract cluster metadata management into a separate module, simplifying cluster module operations.

# Are there any user-facing changes?

# How does this change test
Old tests have covered.

### CODE REVIEW

File: server/coordinator/procedure/scatter.go [hunk 1]
```diff
@@ -49,39 +49,43 @@ func scatterPrepareCallback(event *fsm.Event) {
 
 	waitForNodesReady(c)
 
-	if c.GetClusterState() == clusterpb.ClusterTopology_STABLE {
+	if c.GetClusterState() == storage.Stable {
```
Comment (ZuLiangWang) [inline server/coordinator/procedure/scatter.go:52]: The meaning of this variable `storage.Stable` is not clear, change it to `storage.ClusterState_State`

File: server/coordinator/procedure/create_table.go [hunk 2]
```diff
@@ -42,63 +42,66 @@ var (
...
 		return
 	}
 	// TODO: consider followers
-	leader := shard.GetLeader()
-	if leader == nil {
-		cancelEventWithLog(event, ErrShardLeaderNotFound, "shard can't find leader")
+	leader := storage.ShardNode{}
+	found := false
+	for _, shardNode := range shardNodes {
+		if shardNode.ShardRole == storage.Leader {
```
Comment (ZuLiangWang) [inline server/coordinator/procedure/create_table.go:71]: The meaning of this variable `storage.Leader`  is not clear, change it to `storage.ShardRole_Leader`

File: server/coordinator/scheduler.go [hunk 3]
```diff
@@ -83,27 +84,31 @@ func (s *Scheduler) checkNode(ctx context.Context, ticker *time.Ticker) {
 				log.Error("get node shards failed", zap.Error(err))
 				continue
 			}
-			nodeShardsMapping := map[string][]*cluster.ShardInfo{}
+			nodeShardsMapping := map[string][]cluster.ShardInfo{}
 			for _, nodeShard := range nodeShards.NodeShards {
-				_, exists := nodeShardsMapping[nodeShard.Endpoint]
+				_, exists := nodeShardsMapping[nodeShard.ShardNode.Node]
```
Comment (ZuLiangWang) [inline server/coordinator/scheduler.go:89]:
```suggestion
				_, exists := nodeShardsMapping[nodeShard.ShardNode.NodeName]
```

File: server/coordinator/scheduler.go [hunk 4]
```diff
@@ -83,27 +84,31 @@ func (s *Scheduler) checkNode(ctx context.Context, ticker *time.Ticker) {
...
-func (s *Scheduler) processNodes(ctx context.Context, nodes []*cluster.RegisteredNode, t time.Time, nodeShardsMapping map[string][]*cluster.ShardInfo) {
+func (s *Scheduler) processNodes(ctx context.Context, nodes []cluster.RegisteredNode, t time.Time, nodeShardsMapping map[string][]cluster.ShardInfo) {
 	for _, node := range nodes {
 		// Determines whether node is expired.
 		if !node.IsExpired(uint64(t.Unix()), heartbeatKeepAliveIntervalSec) {
 			// Shard versions of CeresDB and CeresMeta may be inconsistent. And close extra shards and open missing shards if so.
 			realShards := node.GetShardInfos()
-			expectShards := nodeShardsMapping[node.GetMeta().GetName()]
-			err := s.applyMetadataShardInfo(ctx, node.GetMeta().GetName(), realShards, expectShards)
+			expectShards := nodeShardsMapping[node.GetNode().Name]
```
Comment (ZuLiangWang) [inline server/coordinator/scheduler.go:110]:
```suggestion
			expectShards := nodeShardsMapping[node.GetNode().NodeName]
```

File: server/service/grpc/service.go [hunk 5]
```diff
@@ -300,16 +319,16 @@ func convertToGetTablesOfShardsResponse(shardTables map[uint32]*cluster.ShardTab
...
 
-func convertRouteTableResult(routeTablesResult *cluster.RouteTablesResult) *metaservicepb.RouteTablesResponse {
+func convertRouteTableResult(routeTablesResult cluster.RouteTablesResult) *metaservicepb.RouteTablesResponse {
 	entries := make(map[string]*metaservicepb.RouteEntry, len(routeTablesResult.RouteEntries))
 	for tableName, entry := range routeTablesResult.RouteEntries {
 		nodeShards := make([]*metaservicepb.NodeShard, 0, len(entry.NodeShards))
 		for _, nodeShard := range entry.NodeShards {
 			nodeShards = append(nodeShards, &metaservicepb.NodeShard{
-				Endpoint: nodeShard.Endpoint,
+				Endpoint: nodeShard.ShardNode.Node,
```
Comment (ZuLiangWang) [inline server/service/grpc/service.go:328]:
```suggestion
				NodeName: nodeShard.ShardNode.NodeName,
```

**Review by ZuLiangWang — CHANGES_REQUESTED** [review]

File: server/coordinator/scheduler.go (hunk 4 above)
Comment (chunshao90) [reply server/coordinator/scheduler.go:110]: The structure is `Node`, and the name in `Node` do not need Node prefix.

File: server/cluster/types.go [hunk 6]
```diff
@@ -5,90 +5,87 @@ package cluster
...
-type ShardsOfNode struct {
-	Endpoint string
-	ShardIDs []uint32
-}
-
-type NodeShard struct {
-	Endpoint  string
-	ShardInfo *ShardInfo
+type ShardNodeWithVersion struct {
+	version   uint64
```
[the remark this answers is not in the document]
Comment (chunshao90) [reply server/cluster/types.go:34]: `ShardNode` is the memory structure corresponding to `clusterpb.ShardNode`.

**Review by ZuLiangWang — APPROVED** [review]
LGTM

**Pull request merged**